rfcs/2022-07-28-13691-registered-internal-events.md

```diff
@@ -0,0 +1,231 @@
+# RFC 13691 - 2022-07-28 - Registered Internal Events
+
+Emission of named metrics, complete with all the required tags, is a relatively expensive process
+both in terms of generating the metric data on the Vector side, and then determining which global
+metric to update the new data to on the `metrics` crate side. Recent upgrades to the `metrics` crate
+provide the opportunity to register a metric handle once and then update the data for that handle at
+a later point. That update can frequently then be reduced to an indirect jump and atomic memory
+operation. The `InternalEvent` trait, however, does not provide any way to split registration of an
+internal event from the emission of the of that event.  This document proposes an approach to
+solving that problem that can then be applied incrementally to the most expensive internal metrics.
+
+## Context
+
+- Custom registered metric in the `SourceSender` that should be an `InternalEvent`:
+  https://github.com/vectordotdev/vector/pull/13611
+- Performance boost from emiiting events once per batch:
+  https://github.com/vectordotdev/vector/pull/13739
+
+## Scope
+
+### In scope
+
+- Creation of a base interface to standardize registering handles to metrics.
+- Creation of registered events to replace key "hot" metrics.
+- Incremental conversion of hot events to registered events.
+
+### Out of scope
+
+- Conversion of all existing internal events into registered events.
+
+## Pain
+
+- Emission of internal metrics is computationally expensive for hot events.
+
+## Proposal
+
+### User Experience
+
+As an internal event mechanism, this should have no user impact besides making Vector more efficient
+at recording the same internal observability data.
+
+### Implementation
+
+The current setup of internal events using a trait method on structures has the desirable property
+that all parameters are named:
+
+```rust
+BytesSent {
+    byte_size: 12345,
+    protocol: "https",
+}
+.emit();
+```
+
+We include a common wrapper to make it look more functional. The function is also wrapped in a macro
+which is conveniently imported everywhere in the main vector library (and coincidentally allows us
+to do a little magic with event names in tests).
+
+```rust
+emit!(BytesSent {
+    byte_size: 12345,
+    protocol: "https",
+});
+```
+
+To continue the same pattern, we will set up a new trait to register an event. As above, it will
+consume the event data (and so take ownership of all the fields) and return the handle. This new
+return type for the event handle will necessarily be named by the trait.
+
+```rust
+trait RegisterInternalEvent: Sized {
+    type Handle: InternalEventHandle;
```

Comment:
As a struct field, you could either name the actual handle (requiring a bit of inside knowledge) or use something like `<RegisteredByteSize as RegisterInternalEvents>::Handle`. It's fugly, but I don't see a better alternative.

Reply:
Yeah, fair re: just naming the handle type itself.

Another potential thought: a helper type called, say, `Registered<T>`, that you get back from the register call where `T` is the actual registration type? It could still wire itself up all the way through, allowing a blanket impl of `InternalEventHandle` to be added, but then the user would only need to remember the type that they used for registration.

Definitely trading one thing for another, but I'm mostly trying to think of ways where we could reduce the necessary types/methods one needs to be aware of when using this approach.

Reply:
Adding a type alias `type Registered<T> = <T as RegisterInternalEvents>::Handle;` helps _slightly_. Were you thinking of something else?

Reply:
Nah, that's good enough as-is.

Any further improvements would likely involve some sort of macro to do the codegen lifting, and that's obviously not required here to achieve the core design goal.